src/rtltcp.c

```diff
@@ -0,0 +1,193 @@
+#else
+    close(st->socket);
+#endif
+    free(st);
+}
+
+int rtltcp_read(rtltcp_t *st, uint8_t *buf, size_t cnt)
+{
+    int offset = 0;
+    while (cnt > 0)
+    {
+        int err = recv(st->socket, (char *)buf + offset, cnt, 0);
+        if (err < 0)
+            return err;
+        else if (err == 0)
+            break;
+
+        cnt -= err;
+        offset += err;
+    }
+    return offset;
+}
+
+// from https://github.com/steve-m/librtlsdr/blob/master/src/librtlsdr.c
+int rtltcp_get_tuner_gains(rtltcp_t *st, int *gains)
+{
+    /* all gain values are expressed in tenths of a dB */
+    const int e4k_gains[] = { -10, 15, 40, 65, 90, 115, 140, 165, 190, 215,
+        240, 290, 340, 420 };
+    const int fc0012_gains[] = { -99, -40, 71, 179, 192 };
+    const int fc0013_gains[] = { -99, -73, -65, -63, -60, -58, -54, 58, 61,
+        63, 65, 67, 68, 70, 71, 179, 181, 182,
+        184, 186, 188, 191, 197 };
+    const int fc2580_gains[] = { 0 /* no gain values */ };
+    const int r82xx_gains[] = { 0, 9, 14, 27, 37, 77, 87, 125, 144, 157,
+        166, 197, 207, 229, 254, 280, 297, 328,
+        338, 364, 372, 386, 402, 421, 434, 439,
+        445, 480, 496 };
+    const int unknown_gains[] = { 0 /* no gain values */ };
+
+    const int *ptr = NULL;
+    int len = 0;
+
+    switch (st->tuner_type) {
+        case RTLSDR_TUNER_E4000:
+            ptr = e4k_gains; len = sizeof(e4k_gains);
+            break;
+        case RTLSDR_TUNER_FC0012:
+            ptr = fc0012_gains; len = sizeof(fc0012_gains);
+            break;
+        case RTLSDR_TUNER_FC0013:
+            ptr = fc0013_gains; len = sizeof(fc0013_gains);
+            break;
+        case RTLSDR_TUNER_FC2580:
+            ptr = fc2580_gains; len = sizeof(fc2580_gains);
+            break;
+        case RTLSDR_TUNER_R820T:
+        case RTLSDR_TUNER_R828D:
+            ptr = r82xx_gains; len = sizeof(r82xx_gains);
+            break;
+        default:
+            log_error("Unknown tuner type: %d", st->tuner_type);
+            ptr = unknown_gains; len = sizeof(unknown_gains);
+            break;
+    }
+
+    if (!gains)
+    {
+        /* no buffer provided, just return the count */
+        return len / sizeof(int);
+    }
+    else
+    {
+        if (len)
+            memcpy(gains, ptr, len);
+
+        return len / sizeof(int);
+    }
+}
+
+int rtltcp_reset_buffer(rtltcp_t *st, size_t cnt)
+{
+    char buf[1024];
+    unsigned int recvd = 0;
+    int flags = 0;
+#ifdef __MINGW32__
+    unsigned long mode = 1;
+    ioctlsocket(st->socket, FIONBIO, &mode);
+#else
+    flags |= MSG_DONTWAIT;
+#endif
+    // first, clear pending data
+    while (1)
+    {
+        int err = recv(st->socket, buf, sizeof(buf), flags);
+        if (err <= 0)
+            break;
+        recvd += err;
+    }
+#ifdef __MINGW32__
+    mode = 0;
+    ioctlsocket(st->socket, FIONBIO, &mode);
+#endif
+    // if we read an odd-number of bytes, read one more byte
+    recv(st->socket, buf, 1, 0);
```

Comment:
Is this meant to be inside an `if`?

Reply:
Ah, yes. I'll fix that and make sure it still works.